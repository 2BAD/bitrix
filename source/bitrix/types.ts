import { Diff } from './utils/Diff'

// @todo We'd want to have gettable and listable as separate enums,
//       but we can't due too types compatibility issues when `list` uses `get` internally
export enum APIMethod {
  // Gettable
  BATCH = 'batch',
  GET_DEAL = 'crm.deal.get',
  GET_LEAD = 'crm.lead.get',

  // Listable
  LIST_DEALS = 'crm.deal.list',
  LIST_LEADS = 'crm.lead.list'
}

const BITRIX_LISTABLE_METHODS = [APIMethod.LIST_DEALS, APIMethod.LIST_LEADS] as const

export type BitrixListableMethod = typeof BITRIX_LISTABLE_METHODS[number]
export type BitrixGettableMethod = Diff<APIMethod, BitrixListableMethod>

export interface BitrixPayloadTime {
  readonly start: number
  readonly finish: number
  readonly duration: number
  readonly processing: number
  readonly date_start: string
  readonly date_finish: string
}

export interface BitrixGetPayload<P> {
  readonly result: P,
  readonly time: BitrixPayloadTime
}

export interface BitrixListPayload<P> {
  readonly result: readonly P[],
  readonly error?: string,
  readonly total: number,
  readonly next?: number
  readonly time: BitrixPayloadTime
}

// `C` stands for a map of names to structural types they will uphold in result
// `[]` in language of ill Bitrix means `undefined`. Just deal with it.
// Also, it will return object if command names specified and array if names are numbers. Deal with it.
export interface BitrixBatchPayload<C> {
  readonly result: {
    readonly result: { readonly [P in keyof C]?: C[P] } | ReadonlyArray<C[keyof C]>
    readonly result_error: { readonly [P in keyof C]?: string } | readonly string[]
    readonly result_total: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_next: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_time: { readonly [P in keyof C]?: BitrixPayloadTime } | readonly BitrixPayloadTime[]
  }
  readonly time: BitrixPayloadTime
}

export interface BitrixCommandParams {
  readonly [key: string]: string | number
}

export interface BitrixCommand {
  readonly method: APIMethod
  readonly params?: BitrixCommandParams
}

export interface BitrixCommands {
  readonly [key: string]: BitrixCommand
}

export interface BitrixListOptions {
  readonly start?: number
}

// =====================================================
// Bitrix Payloads
// =====================================================

// String which is actually a number, like `'20.23'`
export type NumberString = string
// Like `'2018-06-07T03:00:00+03:00'`
export type ISODate = string
export type BitrixBoolString = 'Y' | 'N'

// @todo This is approximate structure. Might not be accurate
export interface BitrixDeal {
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: unknown // @todo Check is it right
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BitrixBoolString
  readonly CLOSED: BitrixBoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: unknown // @todo Check is it right
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BitrixBoolString
  readonly IS_RECURRING: BitrixBoolString
  readonly IS_RETURN_CUSTOMER: BitrixBoolString
  readonly IS_REPEATED_APPROACH: BitrixBoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
// @todo This is approximate structure. Might not be accurate
export interface BitrixLead {
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BitrixBoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BitrixBoolString
  readonly HAS_EMAIL: BitrixBoolString
  readonly HAS_IMOL: BitrixBoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BitrixBoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
