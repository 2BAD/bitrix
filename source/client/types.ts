import { Diff } from '../utils/Diff'

export enum Method {
  // Gettable
  BATCH = 'batch',

  GET_CONTACT = 'crm.contact.get',
  GET_DEAL = 'crm.deal.get',
  GET_LEAD = 'crm.lead.get',
  GET_STATUS = 'crm.status.get',
  GET_USER = 'user.get',

  CREATE_CONTACT = 'crm.contact.add',
  CREATE_DEAL = 'crm.deal.add',
  CREATE_LEAD = 'crm.lead.add',
  CREATE_STATUS = 'crm.status.add',

  UPDATE_CONTACT = 'crm.contact.update',
  UPDATE_DEAL = 'crm.deal.update',
  UPDATE_LEAD = 'crm.lead.update',
  UPDATE_STATUS = 'crm.status.update',

  // Listable
  LIST_CONTACTS = 'crm.contact.list',
  LIST_DEALS = 'crm.deal.list',
  LIST_LEADS = 'crm.lead.list',
  LIST_STATUSES = 'crm.status.list',
  // yes, this one is correct, they don't have separate `list` method and this one returns all users
  LIST_USERS = 'user.search'
}

const LISTABLE_METHODS = [
  Method.LIST_CONTACTS,
  Method.LIST_DEALS,
  Method.LIST_LEADS,
  Method.LIST_STATUSES,
  Method.LIST_USERS
] as const

export type ListableMethod = typeof LISTABLE_METHODS[number]
export type GettableMethod = Diff<Method, ListableMethod>

export interface PayloadTime {
  readonly start: number
  readonly finish: number
  readonly duration: number
  readonly processing: number
  readonly date_start: string
  readonly date_finish: string
}

export interface GetPayload<P> {
  readonly result: P,
  readonly time: PayloadTime
}

export interface ListPayload<P> {
  readonly result: readonly P[],
  readonly error?: string,
  readonly total: number,
  readonly next?: number
  readonly time: PayloadTime
}

// `C` stands for a map of names to structural types they will uphold in result
// `[]` in language of ill Bitrix means `undefined`. Just deal with it.
// Also, it will return object if command names specified and array if names are numbers. Deal with it.
export interface BatchPayload<C> {
  readonly result: {
    readonly result: { readonly [P in keyof C]?: C[P] } | ReadonlyArray<C[keyof C]>
    readonly result_error: { readonly [P in keyof C]?: string } | readonly string[]
    readonly result_total: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_next: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_time: { readonly [P in keyof C]?: PayloadTime } | readonly PayloadTime[]
  }
  readonly time: PayloadTime
}

export type Payload<P> = GetPayload<P> | ListPayload<P> | BatchPayload<P>

// @todo Figure out full list of possible values
export interface ListParams {
  readonly start?: number
  readonly order?: { readonly [key: string]: 'ASC' }
  readonly filter?: { readonly '>PROBABILITY': number }
  readonly select?: ReadonlyArray<'*' | 'UF_*' | string>
}

export interface Command {
  readonly method: Method
  readonly params?: Record<string, any>
}

export type Commands =
  { readonly [key: string]: Command } |
  // For arrays. It's signature, since `Command[]` won't be
  // accepted by types like `Record`
  { readonly [index: number]: Command }
