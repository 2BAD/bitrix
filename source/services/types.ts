// =====================================================
// Bitrix Payloads
// =====================================================

// String which is actually a number, like `'20.23'`
export type NumberString = string
// Like `'2018-06-07T03:00:00+03:00'`
export type ISODate = string
export type BoolString = 'Y' | 'N'

// @todo This is approximate structure. Might not be accurate
export interface Deal {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: unknown // @todo Check is it right
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BoolString
  readonly CLOSED: BoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: unknown // @todo Check is it right
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BoolString
  readonly IS_RECURRING: BoolString
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly IS_REPEATED_APPROACH: BoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}

// @todo This is approximate structure. Might not be accurate
export interface Lead {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BoolString
  readonly HAS_EMAIL: BoolString
  readonly HAS_IMOL: BoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}

export interface StatusExtra {
  readonly SEMANTICS: string
  readonly COLOR: string
}

export interface Status {
  readonly ID: NumberString
  readonly ENTITY_ID: string
  readonly STATUS_ID: NumberString
  readonly NAME: string
  readonly NAME_INIT: string
  readonly SORT: NumberString
  readonly SYSTEM: BoolString
  readonly EXTRA: StatusExtra | undefined
}
