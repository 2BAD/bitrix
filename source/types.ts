import { BatchPayload, Commands, GetPayload, ListParams, ListPayload, Method } from './client/types'
import { Contact } from './services/types/contact'
import { Deal } from './services/types/deal'
import { Lead } from './services/types/lead'
import { Status } from './services/types/status'
import { ExtractValue } from './utils/ExtractValue'

interface MethodsMap {
  readonly [key: string]: {
    readonly type: unknown
    readonly payload: unknown
    readonly params: Record<string, any>
  }
}

/**
 * A mega map of all supported Bitrix methods to their parameters. Used to resolve
 * low-level client method required params and payload types.
 * - `type` — a type that method associated with
 * - `payload` — a payload that method returns
 * - `params` — params that method accepts
 */
export interface Methods extends MethodsMap {
  readonly [Method.BATCH]: {
    readonly type: unknown
    readonly payload: BatchPayload<unknown>
    readonly params: Commands
  }

  // Deals

  readonly [Method.CREATE_DEAL]: {
    readonly type: Deal
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Deal>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.GET_DEAL]: {
    readonly type: Deal
    readonly payload: GetPayload<Deal>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.UPDATE_DEAL]: {
    readonly type: Deal
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, any>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.LIST_DEALS]: {
    readonly type: Deal
    readonly payload: ListPayload<Deal>
    readonly params: ListParams
  }

  // Leads

  readonly [Method.CREATE_LEAD]: {
    readonly type: Lead
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Lead>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.GET_LEAD]: {
    readonly type: Lead
    readonly payload: GetPayload<Lead>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.LIST_LEADS]: {
    readonly type: Lead
    readonly payload: ListPayload<Lead>
    readonly params: ListParams
  }
  readonly [Method.UPDATE_LEAD]: {
    readonly type: Lead
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, any>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }

  // Contacts

  readonly [Method.CREATE_CONTACT]: {
    readonly type: Contact
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Contact>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.GET_CONTACT]: {
    readonly type: Contact
    readonly payload: GetPayload<Contact>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.LIST_CONTACTS]: {
    readonly type: Contact
    readonly payload: ListPayload<Contact>
    readonly params: ListParams
  }
  readonly [Method.UPDATE_CONTACT]: {
    readonly type: Contact
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, any>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }

  // Statuses

  readonly [Method.CREATE_STATUS]: {
    readonly type: Status
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Status>
    }
  }
  readonly [Method.GET_STATUS]: {
    readonly type: Status
    readonly payload: GetPayload<Status>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.UPDATE_STATUS]: {
    readonly type: Status
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, any>
    }
  }
  readonly [Method.LIST_STATUSES]: {
    readonly type: Status
    readonly payload: ListPayload<Status>
    readonly params: ListParams
  }
}

/**
 * Retrieve specified method data
 */
export type MethodData<M extends Method> = ExtractValue<Methods, M>

/**
 * Retrieve associated with specified method type
 */
export type MethodPayloadType<M extends Method> = MethodData<M>['type']

/**
 * Retrieve specified method payload type
 */
export type MethodPayload<M extends Method> = MethodData<M>['payload']

/**
 * Retrieve specified method params
 */
export type MethodParams<M extends Method> = MethodData<M>['params']
