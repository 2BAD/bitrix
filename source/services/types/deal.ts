import { BoolString, ISODate, NumberString } from './common'

export interface Deal {
  // Deal can have user fields
  readonly [key: string]: string | null

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: NumberString | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BoolString
  readonly CLOSED: BoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: NumberString | null
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BoolString
  readonly IS_RECURRING: BoolString
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly IS_REPEATED_APPROACH: BoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
