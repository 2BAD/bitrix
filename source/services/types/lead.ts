import { BoolString, ISODate, NumberString } from './common'

// @todo This is approximate structure. Might not be accurate
export interface Lead {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BoolString
  readonly HAS_EMAIL: BoolString
  readonly HAS_IMOL: BoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
