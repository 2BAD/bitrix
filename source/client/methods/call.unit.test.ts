/* tslint-env jest */
// tslint:disable: no-expression-statement

import got from 'got'
import nock from 'nock'
import { BatchPayload, GetPayload, ListPayload, Method } from '../types'
import Call, { handlePayload } from './call'

const queueAdd = jest.fn((fn: any) => Promise.resolve(fn()))
const TEST_URI = 'https://test.com/rest'
const call = Call({
  client: got.extend({ baseUrl: TEST_URI, json: true }),
  queue: {
    add: queueAdd
  } as any
})
const RESPONSE_200 = 200

describe('Client `handlePayload` method', () => {
  it('should return `GetPayload`', () => {
    const payload: GetPayload<any> = {
      result: 'done',
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    expect(handlePayload(payload)).toBe(payload)
  })

  it('should return `ListPayload`', () => {
    const payload: ListPayload<any> = {
      error: undefined,
      next: undefined,
      result: ['done'],
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' },
      total: 1
    }

    expect(handlePayload(payload)).toBe(payload)
  })

  it('should throw when getting error in `ListPayload`', () => {
    const payload: ListPayload<any> = {
      error: 'Expected error for `handlePayload`',
      next: undefined,
      result: ['done'],
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' },
      total: 1
    }

    expect(() => handlePayload(payload)).toThrowErrorMatchingSnapshot()
  })

  it('should return `BatchPayload` with numbered commands', () => {
    const payload: BatchPayload<any> = {
      result: {
        result: { one: 'done' },
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    expect(handlePayload(payload)).toBe(payload)
  })

  it('should return `BatchPayload` with result from array of commands', () => {
    const payload: BatchPayload<any> = {
      result: {
        result: ['done'],
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    expect(handlePayload(payload)).toBe(payload)
  })

  it('should throw when getting errors in `BatchPayload` with result from named commands', () => {
    const payload: BatchPayload<any> = {
      result: {
        result: { one: 'done' },
        result_error: { one: 'Expected error for `handlePayload`' },
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    expect(() => handlePayload(payload)).toThrowErrorMatchingSnapshot()
  })

  it('should throw when getting errors in `BatchPayload` with result from array of commands', () => {
    const payload: BatchPayload<any> = {
      result: {
        result: ['done'],
        result_error: ['Expected error for `handlePayload`'],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    expect(() => handlePayload(payload)).toThrowErrorMatchingSnapshot()
  })
})

describe('Client `call` method', () => {
  beforeEach(() => {
    queueAdd.mockClear()
  })

  it('should form a proper request', async () => {
    const params = { id: '1' }

    const scope = nock(TEST_URI)
      .get(`/${Method.GET_DEAL}`)
      .query(params)
      .reply(RESPONSE_200)

    await call(Method.GET_DEAL, params)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper request with list methods', async () => {
    const params = {
      filter: { '>PROBABILITY': 50 },
      order: { STAGE_ID: 'ASC' },
      select: ['ID', 'TITLE'],
      start: 50
    } as const

    const scope = nock(TEST_URI)
      .get(`/${Method.LIST_DEALS}`)
      .query(params)
      .reply(RESPONSE_200)

    await call(Method.LIST_DEALS, params)

    expect(scope.done()).toBe(undefined)
  })

  it('should return body as payload', async () => {
    const params = { id: '1' }
    const payload = { test: 'value' }

    nock(TEST_URI)
      .get(`/${Method.GET_DEAL}`)
      .query(params)
      .reply(RESPONSE_200, payload)

    expect(await call(Method.GET_DEAL, params)).toEqual(payload)
  })

  it('should return body as payload with list methods', async () => {
    const payload = { result: ['done'] }

    nock(TEST_URI)
      .get(`/${Method.LIST_DEALS}`)
      .reply(RESPONSE_200, payload)

    expect(await call(Method.LIST_DEALS, {})).toEqual(payload)
  })

  it('should throw when getting errors in payload', () => {
    const payload = { result: ['done'], error: 'Expected error for `getList`' }

    nock(TEST_URI)
      .get(`/${Method.LIST_DEALS}`)
      .reply(RESPONSE_200, payload)

    return expect(call(Method.LIST_DEALS, {})).rejects.toMatchSnapshot()
  })

  it('should throw when getting errors in batch payload', () => {
    const payload: BatchPayload<any> = {
      result: {
        result: ['done'],
        result_error: ['Expected error'],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: { start: 1, finish: 1, duration: 1, processing: 1, date_start: 'date', date_finish: 'date' }
    }

    nock(TEST_URI)
      .get(`/${Method.BATCH}`)
      .reply(RESPONSE_200, payload)

    return expect(call(Method.BATCH, {})).rejects.toMatchSnapshot()
  })

  it.todo('should cast payload to the <P>')

  it('should add to the queue', async () => {
    const payload = {}
    const method = Method.LIST_DEALS

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${method}`)
      .reply(RESPONSE_200, payload)
      .get(`/${method}`)
      .reply(RESPONSE_200, payload)

    await call(method, {})
    await call(method, {})

    expect(queueAdd.mock.calls).toMatchSnapshot()
  })
})
