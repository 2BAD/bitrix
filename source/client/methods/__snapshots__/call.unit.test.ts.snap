// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client \`call\` method should add to the queue 1`] = `
Array [
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;

exports[`Client \`call\` method should throw when getting errors in batch payload 1`] = `
[Error: [batch] failed to process. Received errors in 1 commands:
Expected error]
`;

exports[`Client \`call\` method should throw when getting errors in payload 1`] = `[Error: [call] failed to get the resource: Expected error for \`getList\`.]`;

exports[`Client \`handlePayload\` method should throw when getting error in \`ListPayload\` 1`] = `"[call] failed to get the resource: Expected error for \`handlePayload\`."`;

exports[`Client \`handlePayload\` method should throw when getting errors in \`BatchPayload\` with result from array of commands 1`] = `
"[batch] failed to process. Received errors in 1 commands:
Expected error for \`handlePayload\`"
`;

exports[`Client \`handlePayload\` method should throw when getting errors in \`BatchPayload\` with result from named commands 1`] = `
"[batch] failed to process. Received errors in 1 commands:
Expected error for \`handlePayload\`"
`;
